cli/flox/src/config/mod.rs

```diff
@@ -189,132 +186,164 @@ pub enum ReadWriteError {
-                builder =
-                    builder.add_source(config::File::from(file).format(config::FileFormat::Toml));
+            fs::create_dir_all(&config_dir)
+                .context(format!("Could not create config directory: {config_dir:?}"))?;
+            let config_dir = config_dir
+                .canonicalize()
+                .context("Could not canonicalize config directory '{config_dir:?}'")?;
+
+            // Allow subshells to find the same config dir.
+            // TODO: decide if its worth modifying the env for this.
+            // SAFTEY: config initially read when there is no concurrent access to env variables.
+            unsafe {
+                env::set_var(FLOX_CONFIG_DIR_VAR, &config_dir);
             }
+            config_dir
+        },
+    };
 
-            // Add explicit FLOX_CONFIG_DIR file last
-            builder = builder.add_source(
-                config::File::from(config_dir.join(FLOX_CONFIG_FILE))
-                    .format(config::FileFormat::Toml)
-                    .required(false),
-            );
-
-            // override via env variables
-            let mut flox_envs = env::vars()
-                .filter_map(|(k, v)| k.strip_prefix("FLOX_").map(|k| (k.to_owned(), v)))
-                .collect::<Vec<_>>();
-
-            let builder = builder
-                .add_source(mk_environment(&mut flox_envs, "NIX"))
-                .add_source(mk_environment(&mut flox_envs, "GITHUB"))
-                .add_source(mk_environment(&mut flox_envs, "FEATURES"))
-                .add_source(
-                    Environment::default()
-                        .source(Some(HashMap::from_iter(flox_envs)))
-                        .try_parsing(true),
-                );
-
-            let final_config = builder.build()?;
-            Ok(final_config)
-        }
-
-        let instance = INSTANCE.get_or_try_init(|| {
-            // If we are initializing the config for the first time,
-            // we don't need to reload right after
-            reload = false;
-            let config = read_raw_config()?;
-
-            Ok::<_, anyhow::Error>(Mutex::new(config))
-        })?;
-
-        let mut config_guard = instance.lock().expect("config mutex poisoned");
-        if reload {
-            *config_guard = read_raw_config()?;
-        }
-
-        Ok(config_guard.deref().clone())
-    }
+    Ok(config_dir)
+}
 
-    /// Creates a [Config] from the environment and config file
-    ///
-    /// When running in tests, the config is reloaded on every call.
-    pub fn parse() -> Result<Config> {
-        #[cfg(test)]
-        let reload = true;
+/// Reads a [HierarchicalConfig] from an optional system config file,
+/// a user config file and environment variables.
+fn raw_config_from_parts(
+    flox_dirs: &BaseDirectories,
+    user_config_dir: PathBuf,
+    system_config_dir: Option<PathBuf>,
+    env: impl IntoIterator<Item = (String, String)>,
+) -> Result<HierarchicalConfig> {
```

Comment:
praise: feels cleaner now that finding the user config file and system config file are split out ❤️